fedmsg/crypto/x509_ng.py

```diff
@@ -19,256 +19,215 @@
-        config.get('ca_cert_location', default_ca_cert_loc),
+    crl_file = utils.load_remote_cert(
+        config.get('crl_location', 'https://fedoraproject.org/fedmsg/crl.pem'),
+        config.get('crl_cache', '/var/cache/fedmsg/crl.pem'),
+        config.get('crl_cache_expiry', 1800),
+        **config
+    )
+    ca_file = utils.load_remote_cert(
+        config.get('ca_cert_location', 'https://fedoraproject.org/fedmsg/ca.crt'),
         config.get('ca_cert_cache', '/etc/pki/fedmsg/ca.crt'),
         config.get('ca_cert_cache_expiry', 0),
-        **config)
+        **config
+    )
 
-    ctx = m2ext.SSL.Context()
-    ctx.load_verify_locations(cafile=cafile)
-    if not ctx.validate_certificate(cert):
-        return fail("X509 certificate is not valid.")
+    with open(ca_file, 'rb') as fd:
+        ca_certificate = fd.read()
+    with open(crl_file, 'rb') as fd:
+        crl = fd.read()
 
-    # Load and check against the CRL
-    crl = _load_remote_cert(
-        config.get('crl_location', 'https://fedoraproject.org/fedmsg/crl.pem'),
-        config.get('crl_cache', '/var/cache/fedmsg/crl.pem'),
-        config.get('crl_cache_expiry', 1800),
-        **config)
-    crl = M2Crypto.X509.load_crl(crl)
-
-    # FIXME -- We need to check that the CRL is signed by our own CA.
-    # See https://bugzilla.osafoundation.org/show_bug.cgi?id=12954#c2
-    # if not ctx.validate_certificate(crl):
-    #    return fail("X509 CRL is not valid.")
-
-    # FIXME -- we check the CRL, but by doing string comparison ourselves.
-    # This is not what we want to be doing.
-    # There is a patch into M2Crypto to handle this for us.  We should use it
-    # once its integrated upstream.
-    # See https://bugzilla.osafoundation.org/show_bug.cgi?id=12954#c2
-    revoked_serials = [long(line.split(': ')[1].strip(), base=16)
-                       for line in crl.as_text().split('\n')
-                       if 'Serial Number:' in line]
-    if cert.get_serial_number() in revoked_serials:
-        subject = cert.get_subject()
-
-        signer = '(no CN)'
-        if subject.nid.get('CN'):
-            entry = subject.get_entries_by_nid(subject.nid['CN'])[0]
-            if entry:
-                signer = entry.get_data().as_text()
-
-        return fail("X509 cert %r, %r is in the Revocation List (CRL)" % (
-            signer, cert.get_serial_number()))
-
-    # If the cert is good, then test to see if the signature in the messages
-    # matches up with the provided cert.
-    rsa_public = cert.get_pubkey().get_rsa()
-    digest = M2Crypto.EVP.MessageDigest('sha1')
-    digest.update(fedmsg.encoding.dumps(message))
     try:
-        if not rsa_public.verify(digest.digest(), signature):
-            raise M2Crypto.RSA.RSAError("RSA signature failed to validate.")
-    except M2Crypto.RSA.RSAError as e:
-        return fail(str(e))
+        _validate_signing_cert(ca_certificate, certificate, crl)
+    except X509StoreContextError as e:
```

Comment:
I presume this step will catch expired certificates and non-properly signed certificates?